infra/Pos/Diffusion/Subscription/Common.hs

```diff
@@ -57,22 +61,38 @@ data SubscriptionTerminationReason =
 -- giving the reason. Notices will be logged before and after the subscription.
 subscribeTo
     :: forall m. (SubscriptionMode m)
-    => Timer -> SendActions m -> NodeId -> m SubscriptionTerminationReason
-subscribeTo keepAliveTimer sendActions peer = do
+    => Timer
+    -> TVar (Map NodeId SubscriptionStatus)
+    -> MVar Millisecond -- ^ Subscription duration.
+    -> SendActions m
+    -> NodeId
+    -> m SubscriptionTerminationReason
+subscribeTo keepAliveTimer subStatus subDuration sendActions peer = do
+    -- Change subscription status as we begin a new subscription
+    alterPeerSubStatus (Just Subscribing)
     logNotice $ msgSubscribingTo peer
+    subStarted <- liftIO $ getTime Monotonic
     -- 'try' is from safe-exceptions, so it won't catch asyncs.
     outcome <- try $ withConnectionTo sendActions peer $ \_peerData -> NE.fromList
         -- Sort conversations in descending order based on their version so that
         -- the highest available version of the conversation is picked.
         [ Conversation convMsgSubscribe
         , Conversation convMsgSubscribe1
         ]
+    -- Change subscription state
+    alterPeerSubStatus Nothing
+    subEnded <- liftIO $ getTime Monotonic
+    liftIO $ modifyMVar_ subDuration
+        (\x -> return $! max x (timeSpecToMilliseconds $ subEnded - subStarted))
     let reason = either Exceptional (maybe Normal absurd) outcome
     logNotice $ msgSubscriptionTerminated peer reason
     return reason
   where
     convMsgSubscribe :: ConversationActions MsgSubscribe Void m -> m t
     convMsgSubscribe conv = do
+        -- We are now subscribed, in rare cases when the connection will be
+        -- dropped this will result in a missleading subscription state.
+        alterPeerSubStatus (Just Subscribed)
```

Comment:
I wonder if it would make any difference if we did this only after the `send conv MsgSubscribe`. What do you think? Maybe not... if this continuation is being run then we definitely have a functioning TCP socket. The chances that the send would fail are slim... Just thinking aloud.